File: apps/desktop/src/distribos/domain/rules.py
"""Domen qoidalari — buyurtma holati, narx, chegirma, qoldiq, qarzdorlik.

Bu modul **sof**: bazaga ham, tarmoqqa ham tegmaydi. Shuning uchun uni
tez va to'liq test qilish mumkin, va bir xil qoidalar Android'da ham
takrorlanadi (kontrakt testlari bilan solishtiriladi).
"""

from __future__ import annotations

from dataclasses import dataclass
from decimal import ROUND_HALF_UP, Decimal

from distribos.persistence.models import MOVEMENT_SIGN, MovementType, OrderState


class DomainError(Exception):
    """Biznes qoidasi buzildi."""


class IllegalStateTransition(DomainError):
    """Buyurtma holati noqonuniy o'tishga urindi."""


# --- buyurtma holat mashinasi ---------------------------------------------

#: Ruxsat etilgan o'tishlar. Bu yerda yo'q o'tish RAD ETILADI (§13).
ORDER_TRANSITIONS: dict[OrderState, frozenset[OrderState]] = {
    OrderState.DRAFT: frozenset({OrderState.CONFIRMED, OrderState.CANCELLED}),
    OrderState.CONFIRMED: frozenset({OrderState.APPROVED, OrderState.CANCELLED}),
    OrderState.APPROVED: frozenset({OrderState.ALLOCATED, OrderState.CANCELLED}),
    OrderState.ALLOCATED: frozenset({OrderState.PICKED, OrderState.CANCELLED}),
    OrderState.PICKED: frozenset({OrderState.SHIPPED, OrderState.CANCELLED}),
    OrderState.SHIPPED: frozenset({OrderState.DELIVERED}),
    OrderState.DELIVERED: frozenset(
        {OrderState.PARTIALLY_RETURNED, OrderState.RETURNED}
    ),
    OrderState.PARTIALLY_RETURNED: frozenset({OrderState.RETURNED}),
    # Yakuniy holatlar — chiqish yo'q.
    OrderState.RETURNED: frozenset(),
    OrderState.CANCELLED: frozenset(),
}

#: Jo'natilgandan keyin bekor qilib bo'lmaydi — tovar allaqachon ketgan.
TERMINAL_STATES: frozenset[OrderState] = frozenset(
    {OrderState.RETURNED, OrderState.CANCELLED}
)


def can_transition(current: OrderState, target: OrderState) -> bool:
    return target in ORDER_TRANSITIONS.get(current, frozenset())


def assert_transition(current: OrderState, target: OrderState) -> None:
    """Noqonuniy o'tishda `IllegalStateTransition`.

    Bu tekshiruv sinxronizatsiyada ham ishlaydi: peer'dan kelgan
    `ORDER_STATE_CHANGED` noqonuniy bo'lsa, u qo'llanmaydi va konflikt
    sifatida ko'rsatiladi — jimgina qabul qilinmaydi.
    """
    if not can_transition(current, target):
        allowed = sorted(s.value for s in ORDER_TRANSITIONS.get(current, frozenset()))
        raise IllegalStateTransition(
            f"{current.value} -> {target.value} mumkin emas. "
            f"Ruxsat etilgan: {allowed or 'yakuniy holat'}"
        )


def is_terminal(state: OrderState) -> bool:
    return state in TERMINAL_STATES


# --- narx va chegirma -----------------------------------------------------

#: Mijoz toifasi -> mahsulotdagi narx maydoni.
PRICE_TIERS: dict[str, str] = {
    "retail": "retail_price",
    "wholesale": "wholesale_price",
    "agent": "agent_price",
}


def resolve_unit_price(product_prices: dict[str, int], price_tier: str) -> int:
    """Mijoz toifasiga mos narxni tanlaydi (tiyinda).

    Toifa narxi 0 bo'lsa — ulgurjiga, u ham bo'lmasa chakanaga tushadi.
    Narx umuman yo'q bo'lsa xato: 0 so'mga sotishni jimgina ruxsat berish
    moliyaviy teshik ochadi.
    """
    field = PRICE_TIERS.get(price_tier, "wholesale_price")
    for candidate in (field, "wholesale_price", "retail_price"):
        price = product_prices.get(candidate, 0)
        if price > 0:
            return int(price)
    raise DomainError("Mahsulot uchun narx belgilanmagan")


def line_total(
    quantity: Decimal | str, unit_price: int, discount_percent: Decimal | str = 0
) -> int:
    """Qator summasi (tiyinda, butun son).

    Yaxlitlash ROUND_HALF_UP — moliyaviy hisobda odatiy. Butun songa
    yaxlitlanadi, ya'ni tiyindan mayda qism qolmaydi.
    """
    qty = Decimal(str(quantity))
    discount = Decimal(str(discount_percent))
    if qty < 0:
        raise DomainError("Miqdor manfiy bo'lishi mumkin emas")
    if not 0 <= discount <= 100:
        raise DomainError("Chegirma 0-100% oralig'ida bo'lishi kerak")

    gross = qty * Decimal(unit_price)
    net = gross * (Decimal(100) - discount) / Decimal(100)
    return int(net.quantize(Decimal("1"), rounding=ROUND_HALF_UP))


@dataclass(frozen=True, slots=True)
class OrderTotals:
    subtotal: int
    discount_total: int
    total: int


def compute_order_totals(lines: list[dict[str, object]]) -> OrderTotals:
    """Buyurtma summalari. Hammasi tiyinda."""
    subtotal = 0
    total = 0
    for line in lines:
        quantity = Decimal(str(line["quantity"]))
        unit_price = int(line["unit_price"])  # type: ignore[arg-type]
        discount = Decimal(str(line.get("discount_percent", 0)))
        gross = int(
            (quantity * Decimal(unit_price)).quantize(Decimal("1"), rounding=ROUND_HALF_UP)
        )
        subtotal += gross
        total += line_total(quantity, unit_price, discount)
    return OrderTotals(subtotal=subtotal, discount_total=subtotal - total, total=total)


# --- kredit limiti --------------------------------------------------------


@dataclass(frozen=True, slots=True)
class CreditCheck:
    allowed: bool
    current_debt: int
    credit_limit: int
    order_total: int
    reason: str = ""

    @property
    def projected_debt(self) -> int:
        return self.current_debt + self.order_total


def check_credit_limit(current_debt: int, credit_limit: int, order_total: int) -> CreditCheck:
    """Kredit limitini tekshiradi.

    `credit_limit == 0` — limit belgilanmagan, cheklov yo'q (odatiy holat
    kichik savdoda). Manfiy qarz = oldindan to'lov, bu ruxsat etiladi.
    """
    projected = current_debt + order_total
    if credit_limit <= 0:
        return CreditCheck(True, current_debt, credit_limit, order_total)
    if projected > credit_limit:
        return CreditCheck(
            False, current_debt, credit_limit, order_total,
            reason=(
                f"Kredit limiti oshib ketadi: {projected / 100:,.0f} > "
                f"{credit_limit / 100:,.0f}"
            ),
        )
    return CreditCheck(True, current_debt, credit_limit, order_total)


# --- ombor qoldig'i -------------------------------------------------------


def apply_movement(
    quantity_on_hand: Decimal, movement_type: str, quantity: Decimal | str
) -> Decimal:
    """Bitta harakatni qoldiqqa qo'llaydi.

    Qoldiq HECH QACHON to'g'ridan-to'g'ri yozilmaydi — u shu funksiya
    orqali harakatlar ketma-ketligidan hisoblanadi (§13). Shuning uchun
    ikki qurilma bir vaqtda sotsa ham, ikkala harakat ham saqlanadi va
    natija to'g'ri chiqadi (yo'qolgan yangilanish bo'lmaydi).
    """
    sign = MOVEMENT_SIGN.get(movement_type)
    if sign is None:
        raise DomainError(f"Noma'lum harakat turi: {movement_type}")
    return quantity_on_hand + Decimal(str(quantity)) * sign


def compute_stock(movements: list[tuple[str, Decimal | str]]) -> tuple[Decimal, Decimal]:
    """Harakatlar ro'yxatidan `(qoldiq, rezerv)` ni hisoblaydi."""
    on_hand = Decimal(0)
    reserved = Decimal(0)
    for movement_type, quantity in movements:
        amount = Decimal(str(quantity))
        if movement_type == MovementType.RESERVATION:
            reserved += amount
        elif movement_type == MovementType.RELEASE:
            reserved -= amount
        else:
            on_hand = apply_movement(on_hand, movement_type, amount)
    return on_hand, reserved


def available_stock(on_hand: Decimal, reserved: Decimal) -> Decimal:
    """Sotish mumkin bo'lgan miqdor."""
    return on_hand - reserved


def check_stock_available(
    on_hand: Decimal, reserved: Decimal, requested: Decimal | str, *, allow_negative: bool = False
) -> bool:
    """Yetarli qoldiq bormi.

    `allow_negative` — ba'zi korxonalar minusga sotishga ruxsat beradi
    (tovar yo'lda). Bu ATAYLAB yoqiladigan sozlama, sukut bo'yicha yo'q.
    """
    if allow_negative:
        return True
    return available_stock(on_hand, reserved) >= Decimal(str(requested))


# --- qarzdorlik -----------------------------------------------------------


def customer_debt(orders_total: int, payments_total: int, returns_total: int = 0) -> int:
    """Mijoz qarzi (tiyinda). Manfiy qiymat — oldindan to'lov."""
    return orders_total - payments_total - returns_total


def allocate_payment(amount: int, open_orders: list[tuple[str, int]]) -> list[tuple[str, int]]:
    """To'lovni ochiq buyurtmalarga taqsimlaydi (eng eskisidan boshlab).

    `open_orders` — `(order_id, qoldiq_qarz)`, eskisidan yangisiga.
    Ortib qolgan summa taqsimlanmaydi — u avans sifatida qoladi.
    """
    if amount < 0:
        raise DomainError("To'lov summasi manfiy bo'lishi mumkin emas")

    allocations: list[tuple[str, int]] = []
    remaining = amount
    for order_id, outstanding in open_orders:
        if remaining <= 0:
            break
        if outstanding <= 0:
            continue
        applied = min(remaining, outstanding)
        allocations.append((order_id, applied))
        remaining -= applied
    return allocations

File: apps/android/domain/src/main/kotlin/uz/distribos/domain/Rules.kt
package uz.distribos.domain

import java.math.BigDecimal
import java.math.RoundingMode

/**
 * Domen qoidalari — buyurtma holati, narx, chegirma, qoldiq, qarzdorlik.
 *
 * Bu fayl Python `distribos/domain/rules.py` ning KOTLIN NUSXASI va u
 * bilan bir xil natija berishi SHART. Ikki tomon farq qilsa, telefonda
 * hisoblangan buyurtma summasi desktopda boshqacha chiqadi — bu esa
 * sinxronizatsiyada tuzatib bo'lmaydigan xato.
 *
 * Moslikni `RulesParityTest` qulflaydi: u Python generatsiya qilgan
 * vektorlarni o'qib solishtiradi.
 *
 * Pul BUTUN son (tiyin). Suzuvchi nuqta ishlatilmaydi.
 */

/** Buyurtma holati. */
enum class OrderState {
    DRAFT, CONFIRMED, APPROVED, ALLOCATED, PICKED, SHIPPED,
    DELIVERED, PARTIALLY_RETURNED, RETURNED, CANCELLED;

    companion object {
        fun from(value: String): OrderState =
            entries.firstOrNull { it.name == value }
                ?: throw DomainException("Noma'lum buyurtma holati: $value")
    }
}

/** Ombor harakati turi. */
enum class MovementType {
    RECEIPT, SALE, RETURN_IN, RETURN_OUT, TRANSFER_OUT, TRANSFER_IN,
    WRITE_OFF, ADJUSTMENT, RESERVATION, RELEASE;

    companion object {
        fun from(value: String): MovementType =
            entries.firstOrNull { it.name == value }
                ?: throw DomainException("Noma'lum harakat turi: $value")
    }
}

class DomainException(message: String) : Exception(message)

class IllegalStateTransitionException(message: String) : Exception(message)

object OrderRules {

    /** Ruxsat etilgan o'tishlar. Bu yerda yo'q o'tish RAD ETILADI. */
    val TRANSITIONS: Map<OrderState, Set<OrderState>> = mapOf(
        OrderState.DRAFT to setOf(OrderState.CONFIRMED, OrderState.CANCELLED),
        OrderState.CONFIRMED to setOf(OrderState.APPROVED, OrderState.CANCELLED),
        OrderState.APPROVED to setOf(OrderState.ALLOCATED, OrderState.CANCELLED),
        OrderState.ALLOCATED to setOf(OrderState.PICKED, OrderState.CANCELLED),
        OrderState.PICKED to setOf(OrderState.SHIPPED, OrderState.CANCELLED),
        OrderState.SHIPPED to setOf(OrderState.DELIVERED),
        OrderState.DELIVERED to setOf(OrderState.PARTIALLY_RETURNED, OrderState.RETURNED),
        OrderState.PARTIALLY_RETURNED to setOf(OrderState.RETURNED),
        // Yakuniy holatlar — chiqish yo'q.
        OrderState.RETURNED to emptySet(),
        OrderState.CANCELLED to emptySet(),
    )

    val TERMINAL: Set<OrderState> = setOf(OrderState.RETURNED, OrderState.CANCELLED)

    fun canTransition(current: OrderState, target: OrderState): Boolean =
        target in (TRANSITIONS[current] ?: emptySet())

    /**
     * Noqonuniy o'tishda xato.
     *
     * Bu tekshiruv sinxronizatsiyada ham ishlaydi: peer'dan kelgan
     * noqonuniy o'tish qo'llanmaydi va konflikt sifatida ko'rsatiladi.
     */
    fun assertTransition(current: OrderState, target: OrderState) {
        if (!canTransition(current, target)) {
            val allowed = TRANSITIONS[current].orEmpty().map { it.name }.sorted()
            throw IllegalStateTransitionException(
                "${current.name} -> ${target.name} mumkin emas. " +
                    "Ruxsat etilgan: ${allowed.ifEmpty { listOf("yakuniy holat") }}"
            )
        }
    }

    fun isTerminal(state: OrderState): Boolean = state in TERMINAL
}

object PricingRules {

    /** Mijoz toifasi -> mahsulotdagi narx maydoni. */
    val PRICE_TIERS: Map<String, String> = mapOf(
        "retail" to "retail_price",
        "wholesale" to "wholesale_price",
        "agent" to "agent_price",
    )

    /**
     * Mijoz toifasiga mos narxni tanlaydi (tiyinda).
     *
     * Narx umuman yo'q bo'lsa xato: 0 so'mga sotishni jimgina ruxsat
     * berish moliyaviy teshik ochadi.
     */
    fun resolveUnitPrice(prices: Map<String, Long>, priceTier: String): Long {
        val field = PRICE_TIERS[priceTier] ?: "wholesale_price"
        for (candidate in listOf(field, "wholesale_price", "retail_price")) {
            val price = prices[candidate] ?: 0L
            if (price > 0) return price
        }
        throw DomainException("Mahsulot uchun narx belgilanmagan")
    }

    /**
     * Qator summasi (tiyinda, butun son).
     *
     * Yaxlitlash HALF_UP — Python `ROUND_HALF_UP` bilan bir xil.
     */
    fun lineTotal(quantity: String, unitPrice: Long, discountPercent: String = "0"): Long {
        val amount = BigDecimal(quantity)
        val discount = BigDecimal(discountPercent)
        if (amount < BigDecimal.ZERO) throw DomainException("Miqdor manfiy bo'lishi mumkin emas")
        if (discount < BigDecimal.ZERO || discount > BigDecimal(100)) {
            throw DomainException("Chegirma 0-100% oralig'ida bo'lishi kerak")
        }
        val gross = amount.multiply(BigDecimal(unitPrice))
        val net = gross.multiply(BigDecimal(100).subtract(discount))
            .divide(BigDecimal(100), 10, RoundingMode.HALF_UP)
        return net.setScale(0, RoundingMode.HALF_UP).toLong()
    }

    data class OrderTotals(val subtotal: Long, val discountTotal: Long, val total: Long)

    data class LineInput(
        val quantity: String,
        val unitPrice: Long,
        val discountPercent: String = "0",
    )

    fun computeOrderTotals(lines: List<LineInput>): OrderTotals {
        var subtotal = 0L
        var total = 0L
        for (line in lines) {
            val gross = BigDecimal(line.quantity)
                .multiply(BigDecimal(line.unitPrice))
                .setScale(0, RoundingMode.HALF_UP)
                .toLong()
            subtotal += gross
            total += lineTotal(line.quantity, line.unitPrice, line.discountPercent)
        }
        return OrderTotals(subtotal, subtotal - total, total)
    }
}

object CreditRules {

    data class CreditCheck(
        val allowed: Boolean,
        val currentDebt: Long,
        val creditLimit: Long,
        val orderTotal: Long,
        val reason: String = "",
    ) {
        val projectedDebt: Long get() = currentDebt + orderTotal
    }

    /**
     * `creditLimit == 0` — limit belgilanmagan, cheklov yo'q.
     * Manfiy qarz = oldindan to'lov, bu ruxsat etiladi.
     */
    fun check(currentDebt: Long, creditLimit: Long, orderTotal: Long): CreditCheck {
        val projected = currentDebt + orderTotal
        if (creditLimit <= 0) {
            return CreditCheck(true, currentDebt, creditLimit, orderTotal)
        }
        if (projected > creditLimit) {
            return CreditCheck(
                false, currentDebt, creditLimit, orderTotal,
                reason = "Kredit limiti oshib ketadi",
            )
        }
        return CreditCheck(true, currentDebt, creditLimit, orderTotal)
    }
}

object StockRules {

    /** Qaysi harakatlar qoldiqni oshiradi (+1) yoki kamaytiradi (-1). */
    val MOVEMENT_SIGN: Map<MovementType, Int> = mapOf(
        MovementType.RECEIPT to +1,
        MovementType.RETURN_IN to +1,
        MovementType.TRANSFER_IN to +1,
        MovementType.SALE to -1,
        MovementType.RETURN_OUT to -1,
        MovementType.TRANSFER_OUT to -1,
        MovementType.WRITE_OFF to -1,
        MovementType.ADJUSTMENT to +1,   // ishorasi miqdorning o'zida
        MovementType.RESERVATION to 0,
        MovementType.RELEASE to 0,
    )

    data class Stock(val onHand: BigDecimal, val reserved: BigDecimal) {
        val available: BigDecimal get() = onHand.subtract(reserved)
    }

    /**
     * Qoldiq HECH QACHON to'g'ridan-to'g'ri yozilmaydi — u harakatlar
     * ketma-ketligidan hisoblanadi. Shuning uchun ikki qurilma bir vaqtda
     * sotsa ham natija to'g'ri chiqadi (yo'qolgan yangilanish bo'lmaydi).
     */
    fun computeStock(movements: List<Pair<MovementType, String>>): Stock {
        var onHand = BigDecimal.ZERO
        var reserved = BigDecimal.ZERO
        for ((type, quantity) in movements) {
            val amount = BigDecimal(quantity)
            when (type) {
                MovementType.RESERVATION -> reserved = reserved.add(amount)
                MovementType.RELEASE -> reserved = reserved.subtract(amount)
                else -> {
                    val sign = MOVEMENT_SIGN[type]
                        ?: throw DomainException("Noma'lum harakat turi: $type")
                    onHand = onHand.add(amount.multiply(BigDecimal(sign)))
                }
            }
        }
        return Stock(onHand, reserved)
    }

    fun isAvailable(
        stock: Stock, requested: String, allowNegative: Boolean = false,
    ): Boolean = allowNegative || stock.available >= BigDecimal(requested)
}

object FinanceRules {

    /** Mijoz qarzi = buyurtmalar - sof to'lovlar. Manfiy = avans. */
    fun customerDebt(ordersTotal: Long, paymentsTotal: Long, returnsTotal: Long = 0): Long =
        ordersTotal - paymentsTotal - returnsTotal

    /**
     * To'lovni ochiq buyurtmalarga taqsimlaydi (eng eskisidan boshlab).
     * Ortib qolgan summa taqsimlanmaydi — u avans bo'lib qoladi.
     */
    fun allocatePayment(
        amount: Long, openOrders: List<Pair<String, Long>>,
    ): List<Pair<String, Long>> {
        if (amount < 0) throw DomainException("To'lov summasi manfiy bo'lishi mumkin emas")

        val allocations = mutableListOf<Pair<String, Long>>()
        var remaining = amount
        for ((orderId, outstanding) in openOrders) {
            if (remaining <= 0) break
            if (outstanding <= 0) continue
            val applied = minOf(remaining, outstanding)
            allocations += orderId to applied
            remaining -= applied
        }
        return allocations
    }
}
